src/jackdaw/streams/xform.clj

```diff
@@ -0,0 +1,67 @@
+(ns jackdaw.streams.xform
+  "Helper functions for working with transducers."
+  (:gen-class)
+  (:require [jackdaw.serdes :as js]
+            [jackdaw.streams :as j])
+  (:import org.apache.kafka.streams.kstream.Transformer
+           org.apache.kafka.streams.KeyValue
+           [org.apache.kafka.streams.state KeyValueStore Stores]
+           org.apache.kafka.streams.StreamsBuilder))
+
+(defn fake-kv-store
+  "Creates an instance of org.apache.kafka.streams.state.KeyValueStore
+  with overrides for get and put."
+  [init]
+  (let [store (volatile! init)]
+    (reify KeyValueStore
+      (get [_ k]
+        (clojure.core/get @store k))
+
+      (put [_ k v]
+        (vswap! store assoc k v)))))
+
+(defn kv-store-swap-fn
+  "Takes an instance of KeyValueStore, a function f, and map m, and
+  updates the store in a manner similar to `clojure.core/swap!`."
+  [^KeyValueStore store f m]
+  (let [ks (keys (f {} m))
+        prev (reduce (fn [p k]
+                       (assoc p k (.get store k)))
+                     {}
+                     ks)
+        next (f prev m)]
+    (doall (map (fn [[k v]] (.put store k v)) next))
+    next))
+
+(defn add-state-store!
+  [builder]
+  "Takes a builder and adds a state store."
+  (doto ^StreamsBuilder (j/streams-builder* builder)
+    (.addStateStore (Stores/keyValueStoreBuilder
+                     (Stores/persistentKeyValueStore "transducer")
+                     (js/edn-serde)
+                     (js/edn-serde))))
+  builder)
+
+(defn transformer
+  "Takes a transducer and creates an instance of
+  org.apache.kafka.streams.kstream.Transformer with overrides for
+  init, transform, and close."
+  [xf]
+  (let [ctx (atom nil)]
+    (reify
+      Transformer
+      (init [_ context]
+        (reset! ctx context))
+      (transform [_ k v]
+        (let [^KeyValueStore store (.getStateStore @ctx "transducer")
```

Comment:
I would add an airily to pass the backing store name

Reply:
I kind of think this function is the only thing that should really be in this PR (plus a test in xform_test.clj that proves it works). Everything else feels like arbitrary policy decisions that shouldn't be made at the library level.

Reply:
I disagree. `fake-kv-store` and `kv-store-swap-fn` are what make our transducers reusable. We can test our business logic without the `TopologyTestDriver` and Kafka Streams. `kv-store-swap-fn` allows you to write your transducers as if you were using Clojure atoms. You don't have to use it, but if you want it, it's there. The examples show how to use all this to solve non-trivial problems. I think we need them. I could see adding a few tests though.